// @adobe/dc-services-node-sdk/src/internal/config/cpf-engine-config.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

const cpfConfig = {
	createPdf: {
		name: 'jkbm5yryaff',
		invocation_mode: 'asynchronous',
		invocation_batch: false,
		in_response: false,
		engine: 'Classification:sensei-demo:Service-c8a0236bc724493ebe49042a5d0b5370'
	},
	combinePdf: {
		name: '6mszi9yb526',
		invocation_mode: 'asynchronous',
		invocation_batch: false,
		in_response: false,
		engine: 'Classification:sensei-demo:Service-2888f3ebfd5341e0a76b466a61cbe4c6'
	},
	exportPdf: {
		name: 'mhrav60llsr',
		invocation_mode: 'asynchronous',
		invocation_batch: false,
		in_response: false,
		engine: 'Classification:sensei-demo:Service-58e18c3722d744599304bf0b38e61930'
	},
	htmlToPdf: {
		name: 'e5pi3j13mvh',
		invocation_mode: 'asynchronous',
		invocation_batch: false,
		in_response: false,
		engine: 'Classification:sensei-demo:Service-d28a398065794200b941f4a0f00bae19'
	},
	ocr: {
		name: 'ysmpudep2vs',
		invocation_mode: 'asynchronous',
		invocation_batch: false,
		in_response: false,
		engine: 'Classification:sensei-demo:Service-9dedf43251004f968e871fb9f72df46f'
	},

	formatKeyString:"dc:format",
	multipartKeyString:'sensei:multipart_field_name'

};

module.exports = cpfConfig;
