// @adobe/dc-services-node-sdk/src/internal/cpf/request/contentAnalyzerRequests.js
/*
 * Copyright 2019 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. If you have received this file from a source other than Adobe,
 * then your use, modification, or distribution of it requires the prior
 * written permission of Adobe.
 */

class ContentAnaylzerRequests {
	constructor(cpfEngineConfig, engine){
		this['sensei:name'] = cpfEngineConfig.name;
		this['sensei:invocation_mode'] = cpfEngineConfig.invocation_mode;
		this['sensei:invocation_batch'] = cpfEngineConfig.invocation_batch;
		this['sensei:in_response'] = cpfEngineConfig.in_response;
		engine['sensei:execution_info'] = {'sensei:engine': cpfEngineConfig.engine};
		this['sensei:engines'] = [engine];
	}
	setName(value){
		this['sensei:name'] = value;

	}
	setInvocationMode(value){
		this['sensei:invocation_mode'] = value;

	}
	setInvocationBatch(value){
		this['sensei:invocation_batch'] = value;

	}
	setInResponse(value){
		this['sensei:in_response'] = value;

	}
	setEngines(value){
		this['sensei:engines'] = value;

	}
}

module.exports = ContentAnaylzerRequests;
